pkg/discovery/backends.go

```diff
@@ -39,6 +40,70 @@ func parse(rawurl string) (string, string) {
 	return parts[0], parts[1]
 }
 
+// ParseAdvertise parses the --cluster-advertise daemon config which accepts
+// <ip-address>:<port> or <interface-name>:<port>
+func ParseAdvertise(store, advertise string) (string, error) {
+	var (
+		iface *net.Interface
+		addrs []net.Addr
+		err   error
+	)
+
+	addrPort := strings.Split(advertise, ":")
+
+	if len(addrPort) > 2 {
+		return "", fmt.Errorf("cluster advertise config (%s) doesnt match <ipv4-address>:<port> format", advertise)
+	}
+
+	addrStr := addrPort[0]
+	ip := net.ParseIP(addrStr)
+
+	// If advertise is a valid ip-address, use it as is
+	if ip != nil {
+		addrStr = ip.String()
+		if len(addrPort) > 1 {
+			addrStr = fmt.Sprintf("%s:%s", addrStr, addrPort[1])
+		}
+		return addrStr, nil
+	}
+
+	// If advertise is a valid interface name, get the valid ipv4 address and use it to advertise
+	iface, err = net.InterfaceByName(addrStr)
+	if err != nil {
+		return "", fmt.Errorf("invalid cluster advertise ip-address or interface name (%s) : %v", advertise, err)
+	}
+
+	addrs, err = iface.Addrs()
+	if err != nil {
+		return "", fmt.Errorf("error deriving advertise ip-address in interface (%s) : %v", advertise, err)
```

Comment:
s/in/from/ (as in "to derive something *from*")

Alternatively (it's already an "error", so probably no need to start with "error"); something like:

```
unable to get advertise IP-address from interface
```